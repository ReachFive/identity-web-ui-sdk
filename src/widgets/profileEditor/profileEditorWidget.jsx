import React from 'react';

import { UserError } from '../../helpers/errors';
import { snakeCaseProperties, camelCaseProperties } from '../../helpers/transformObjectProperties';

import { createWidget } from '../../components/widget/widget';
import { createForm } from '../../components/form/formComponent';
import { buildFormFields, computeFieldList } from '../../components/form/formFieldFactory';

const ProfileEditorForm = createForm({
    prefix: 'r5-profile-editor-',
    supportMultipleSubmits: true,
    submitLabel: 'save'
});

class ProfileEditor extends React.Component {
    static defaultProps = {
        onSuccess: () => { },
        onError: () => { }
    };

    handleSubmit = data => this.props.apiClient.updateProfile({
        data: snakeCaseProperties(data),
        accessToken: this.props.accessToken,
        redirectUrl: this.props && this.props.redirectUrl
    });

    render() {
        return <ProfileEditorForm
            handler={this.handleSubmit}
            initialModel={this.props.profile}
            fields={this.props.resolvedFields}
            showLabels={this.props.showLabels}
            onSuccess={this.props.onSuccess}
            onError={this.props.onError} />;
    }
}

export default createWidget({
    component: ProfileEditor,
    prepare: (options, { apiClient, config }) => {
        const opts = {
            showLabels: true,
            fields: [],
            ...options
        };
        const { accessToken, fields = [] } = opts;

        const haveNotAllowedFields = fields.some(field => {
            const fieldName = field.key || field;

            return fieldName === 'password' || fieldName === 'password_confirmation'
        });

        if (haveNotAllowedFields) {
            throw new UserError('These fields are not allowed: password, password_confirmation.');
        }

        // This step removes the version from the consents
        const resolvedFields = buildFormFields(fields, { ...config, errorArchivedConsents: false });

        return apiClient
            .getUser({
                accessToken,
                fields: computeFieldList(resolvedFields)
            })
            .then(profile => {
                const camelProfile = camelCaseProperties(profile);
                const profileConsents = camelProfile.consents;
                const filteredProfileConsents = (profileConsents !== undefined && Object.keys(profileConsents).length) ? filterProfileConsents(fields, profileConsents) : undefined;
                const filteredOutConsentsProfile = { ...profile, consents: filteredProfileConsents };
                return ({
                    ...opts,
                    profile: filteredOutConsentsProfile,
                    resolvedFields: resolvedFields.filter(field => {
                        return (field.path !== 'email' || !filteredOutConsentsProfile.email)
                            && (field.path !== 'phone_number' || !config.sms || !filteredOutConsentsProfile.phoneNumber);
                    })
                })
            });
    }
});

// Filter out the profile consents with different version than the one the given consent field own
const filterProfileConsents = (fields, profileConsents) => {
    return Object.keys(profileConsents)
        .filter(profileConsentKey => {
            const consentField = fields.find(field => field.startsWith(`consents.${profileConsentKey}`));
            const consentFieldSplit = consentField.split('.v');
            const consentFieldVersion = consentFieldSplit[1];
            const profileConsentVersion = profileConsents[profileConsentKey].consentVersion.versionId;
            return !consentFieldVersion || parseInt(consentFieldVersion) === profileConsentVersion;
        })
        .reduce((filteredProfileConsents, consentKey) => {
            filteredProfileConsents[consentKey] = profileConsents[consentKey];
            return filteredProfileConsents;
        }, {});
}
