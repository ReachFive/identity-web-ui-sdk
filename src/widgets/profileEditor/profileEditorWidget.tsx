import React from 'react';
import { ConsentVersions, Profile, UserConsent } from '@reachfive/identity-core';

import { UserError } from '../../helpers/errors';
import { camelCaseProperties } from '../../helpers/transformObjectProperties';

import { createWidget } from '../../components/widget/widget';
import { createForm } from '../../components/form/formComponent';
import { buildFormFields, computeFieldList } from '../../components/form/formFieldFactory';
import { useReachfive } from '../../contexts/reachfive';
import { FieldCreator } from '../../components/form/fieldCreator';
import { Field } from '../../components/form/formFieldFactory';
import { PhoneNumberOptions } from '../../components/form/fields/phoneNumberField';

type ProfileWithConsents = Profile & { consents?: Record<string, UserConsent> }

const ProfileEditorForm = createForm<ProfileWithConsents>({
    prefix: 'r5-profile-editor-',
    supportMultipleSubmits: true,
    submitLabel: 'save'
});

interface ProfileEditorProps {
    /**
     * The authorization credential JSON Web Token (JWT) used to access the ReachFive API, less than five minutes old.
     */
    accessToken: string
    /**
     * Callback function called when the request has succeed.
     */
    onSuccess?: () => void
    /**
     * Callback function called when the request has failed.
     */
    onError?: (error?: unknown) => void
    /**
     * Phone number field options.
     */
    phoneNumberOptions?: PhoneNumberOptions
    /**
     * 
     */
    profile: ProfileWithConsents
    /**
     * The URL sent in the email to which the user is redirected. 
     * This URL must be whitelisted in the `Allowed Callback URLs` field of your ReachFive client settings.
     */
    redirectUrl?: string,
    /**
     * 
     */
    resolvedFields: FieldCreator<unknown>[]
    /**
     * Whether the form fields' labels are displayed on the form view.
     * @default false
     */
    showLabels?: boolean
}

const ProfileEditor = ({
    accessToken,
    onSuccess = () => {},
    onError = () => {},
    phoneNumberOptions,
    profile,
    redirectUrl,
    resolvedFields,
    showLabels = false,
}: ProfileEditorProps) => {
    const coreClient = useReachfive()

    const handleSubmit = (data: ProfileWithConsents) =>
        coreClient.updateProfile({
            data,
            accessToken: accessToken,
            redirectUrl: redirectUrl
        });

    return (
        <ProfileEditorForm
            handler={handleSubmit}
            initialModel={profile}
            fields={resolvedFields}
            sharedProps={{
                ...phoneNumberOptions,
            }}
            showLabels={showLabels}
            onSuccess={onSuccess}
            onError={onError}
        />
    )
}

export interface ProfileEditorWidgetProps extends Omit<ProfileEditorProps, 'profile' | 'resolvedFields'> {
    /**
     * List of the fields to display in the form.
     * 
     * **Important:**
     * 
     * The following fields can not be changed with this widget:
     * - `password`
     * - `password_confirmation`
     * 
     * It is not possible to update the primary identifier submitted at registration (email or phone number). When the primary identifier is the email address (SMS feature disabled), users can only enter a phone number and update without limit.
     */
    fields?: (string | Field)[]
}

export default createWidget<ProfileEditorWidgetProps, ProfileEditorProps>({
    component: ProfileEditor,
    prepare: (options, { apiClient, config }) => {
        const opts = {
            showLabels: true,
            fields: [],
            ...options
        };
        const { accessToken, fields = [] } = opts;

        const haveNotAllowedFields = fields.some(field => {
            const fieldName = typeof field === 'string' ? field : field.key;
            return fieldName === 'password' || fieldName === 'password_confirmation'
        });

        if (haveNotAllowedFields) {
            throw new UserError('These fields are not allowed: password, password_confirmation.');
        }

        // This step removes the version from the consents
        const resolvedFields = buildFormFields(fields, { ...config, errorArchivedConsents: false });

        return apiClient
            .getUser({
                accessToken,
                fields: computeFieldList(resolvedFields)
            })
            .then(profile => camelCaseProperties(profile) as Profile) /** @todo check api response key format in sdk core */
            .then((profile: Profile) => {
                const filteredProfileConsents = (profile.consents && Object.keys(profile.consents).length > 0) ? filterProfileConsents(fields, config.consentsVersions, profile.consents) : undefined;
                const filteredOutConsentsProfile = { ...profile, consents: filteredProfileConsents };
                return ({
                    ...opts,
                    profile: filteredOutConsentsProfile,
                    resolvedFields: resolvedFields.filter(field => {
                        return (field.path !== 'email' || !filteredOutConsentsProfile.email)
                            && (field.path !== 'phone_number' || !config.sms || !filteredOutConsentsProfile.phoneNumber);
                    })
                })
            })
            .catch((error: unknown) => {
                options.onError?.(error)
                return Promise.reject(error)
            });
    }
});

// Filter out the profile consents with different version than the one the given consent field own
const filterProfileConsents = (fields: (string | Field)[], consentsVersions: Record<string, ConsentVersions>, profileConsents: Record<string, UserConsent>) => {
    return Object.keys(profileConsents)
        .filter(profileConsentKey => {
            const consentField = fields.map(f => typeof f === 'string' ? f : f.key).find(field => field.startsWith(`consents.${profileConsentKey}`));
            const consentFieldSplit = consentField ? consentField.split('.v') : [];
            // Find most recent consent version if not given
            const highestConsentVersion = consentsVersions[profileConsentKey].versions[0].versionId;
            const consentFieldVersion = parseInt(consentFieldSplit[1]) || highestConsentVersion;
            const profileConsentVersion = profileConsents[profileConsentKey].consentVersion?.versionId;
            return !consentFieldVersion || consentFieldVersion === profileConsentVersion;
        })
        .reduce<Record<string, UserConsent>>((filteredProfileConsents, consentKey) => {
            filteredProfileConsents[consentKey] = profileConsents[consentKey];
            return filteredProfileConsents;
        }, {});
}
