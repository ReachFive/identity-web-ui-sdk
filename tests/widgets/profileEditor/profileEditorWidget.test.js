import renderer from 'react-test-renderer';
import 'jest-styled-components';

import profileEditorWidget from '../../../src/widgets/profileEditor/profileEditorWidget';

const defaultConfig = {
    domain: 'local.reach5.net'
};

describe('Snapshot', () => {
    const generateSnapshot = (options, config = defaultConfig, user) => () => {
        const apiClient = {
            getUser: jest.fn().mockReturnValueOnce(Promise.resolve(user))
        };

        const tree = profileEditorWidget(
            { ...options, accessToken: 'azerty' },
            { config, apiClient }
        ).then(result => renderer.create(result).toJSON());

        expect(tree).resolves.toMatchSnapshot();
    };

    describe('profile editor', () => {
        test('basic', generateSnapshot({
            fields: [
                'given_name',
                'family_name'
            ]
        }));
    });
})
