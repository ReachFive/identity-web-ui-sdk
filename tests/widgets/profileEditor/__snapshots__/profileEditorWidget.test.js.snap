// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot profile editor basic 1`] = `
.c0 {
  font-size: 14px;
  -webkit-transition: -webkit-transform 400ms ease, opacity 400ms ease;
  -webkit-transition: transform 400ms ease, opacity 400ms ease;
  transition: transform 400ms ease, opacity 400ms ease;
  opacity: 0;
  padding: 20px;
  border-radius: 3px;
  background-color: #ffffff;
  max-width: 400px;
  box-sizing: border-box;
  margin: 0 auto;
}

.c5 {
  display: block;
  width: 100%;
  box-sizing: border-box;
  text-align: center;
  font-weight: bold;
  vertical-align: middle;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  touch-action: manipulation;
  cursor: pointer;
  color: #ffffff;
  background-color: #229955;
  border: 1px solid #229955;
  padding: 9px 12px;
  font-size: 14px;
  line-height: 1.428571429;
  border-radius: 3px;
  -webkit-transition: all .15s ease-in-out;
  transition: all .15s ease-in-out;
}

.c5:focus {
  outline: 0;
  box-shadow: 0 0 0 3px rgba(34,153,85,0.5);
}

.c5:hover,
.c5:active {
  background-color: #1b7842;
  border-color: #1b7842;
}

.c5[disabled] {
  opacity: .65;
}

.c1 {
  position: relative;
}

.c3 {
  color: #495057;
  margin-bottom: 5px;
  display: inline-block;
}

.c2 {
  margin-bottom: 10px;
}

.c4 {
  display: block;
  width: 100%;
  box-sizing: border-box;
  font-size: 14px;
  line-height: 1.428571429;
  color: #495057;
  border-radius: 3px;
  border-width: 1px;
  border-style: solid;
  border-color: #ced4da;
  background-color: #fff;
  background-image: none;
  padding: 9px 12px;
  -webkit-transition: border-color ease-in-out .15s, box-shadow ease-in-out .15s;
  transition: border-color ease-in-out .15s, box-shadow ease-in-out .15s;
  box-shadow: none;
  -webkit-appearance: none;
}

.c4:focus {
  border-color: #5fdb94;
  box-shadow: 0 0 0 3px rgba(95,219,148,0.5);
  outline: 0;
}

.c4:disabled,
.c4[readonly] {
  background-color: #e9ecef;
}

.c4::-webkit-input-placeholder {
  color: #868e96;
}

.c4::-moz-placeholder {
  color: #868e96;
}

.c4:-ms-input-placeholder {
  color: #868e96;
}

.c4::placeholder {
  color: #868e96;
}

<div>
  <div
    class="c0"
  >
    <form
      class="c1"
      novalidate=""
    >
      <div
        class="c2"
      >
        <label
          class="c3"
          for="given_name"
        >
          givenName
        </label>
        <input
          class="c4"
          data-testid="given_name"
          id="given_name"
          name="given_name"
          placeholder="givenName"
          required=""
          title="givenName"
          value="John"
        />
      </div>
      <div
        class="c2"
      >
        <label
          class="c3"
          for="family_name"
        >
          familyName
        </label>
        <input
          class="c4"
          data-testid="family_name"
          id="family_name"
          name="family_name"
          placeholder="familyName"
          required=""
          title="familyName"
          value="Do"
        />
      </div>
      <button
        class="c5"
        data-testid="submit"
        type="submit"
      >
        save
      </button>
    </form>
  </div>
</div>
`;
